scanpipe/pipes/windows.py

```diff
@@ -0,0 +1,237 @@
+
+from scanpipe import pipes
+
+
+def package_getter(root_dir, **kwargs):
+    """
+    Yield installed package objects.
+    """
+    packages = win_reg.get_installed_packages(root_dir)
+    for package in packages:
+        yield package.purl, package
+
+
+def tag_uninteresting_windows_codebase_resources(project):
+    """
+    Tag known uninteresting files as uninteresting
+    """
+    uninteresting_files = (
+        "DefaultUser_Delta",
+        "Sam_Delta",
+        "Security_Delta",
+        "Software_Delta",
+        "System_Delta",
+        "NTUSER.DAT",
+        "desktop.ini",
+        "BBI",
+        "BCD-Template",
+        "DEFAULT",
+        "DRIVERS",
+        "ELAM",
+        "SAM",
+        "SECURITY",
+        "SOFTWARE",
+        "SYSTEM",
+    )
+
+    uninteresting_file_extensions = (
+        ".lnk",
+        ".library-ms",
+        ".LOG",
+        ".inf_loc",
+        ".NLS",
+    )
+
+    lookups = Q()
+    for file_name in uninteresting_files:
+        lookups |= Q(path__iendswith=file_name)
+    for file_extension in uninteresting_file_extensions:
+        lookups |= Q(extension__icontains=file_extension)
+
+    qs = project.codebaseresources.no_status()
+    qs.filter(lookups).update(status="ignored-not-interesting")
+
+
+def tag_installed_package_files(project, root_dir_pattern, package):
+    """
+    For all CodebaseResources from `project` whose `rootfs_path` starts with
+    `root_dir_pattern`, add `package` to the discovered_packages of each
+    CodebaseResource and set the status
+    """
+    qs = project.codebaseresources.no_status()
+    installed_package_files = qs.filter(rootfs_path__startswith=root_dir_pattern)
+    # If we find files whose names start with `root_dir_pattern`, we consider
+    # these files to be part of the Package `package` and tag these files as
+    # such
+    if installed_package_files:
+        created_package = pipes.update_or_create_package(
+            project=project, package_data=package.to_dict()
+        )
+        for installed_package_file in installed_package_files:
+            installed_package_file.discovered_packages.add(created_package)
+            installed_package_file.status = "installed-package"
+            installed_package_file.save()
+        created_package.save()
+
+
+def tag_known_software(project):
+    """
+    Find Windows software in `project` by checking `project`s CodebaseResources
+    to see if their rootfs_path is is under a known software root directory. If
+    there are CodebaseResources that are under a known software root directory,
+    a DiscoveredPackage is created for that software package and all files under
+    that software package's root directory are considered installed files for
+    that package.
+
+    Currently, we are only checking for Python and openjdk in Windows Docker
+    image layers.
+    """
+    qs = project.codebaseresources.no_status()
+    python_root_directory_name_pattern = r"/Files/Python(\d*)"
```

Comment:
The `/Files` in the path should be made somewhat configurable such that we can also support non-Docker based rootfs images. This is not for here but everywhere.

Reply:
Regex has been updated such that we are only looking for CodebaseResources that start with `/Python*` or `/Files/Python*`